scripts/install-nix-from-closure.sh

```diff
@@ -1,181 +1,700 @@
-        echo "              installation even when the preferred"
-        echo "              method is with the daemon."
-        echo ""
-    ) >&2
-    exit
-fi
+    Installer is incomplete ($self/.reginfo is missing)
+    "
+        fi
+    }
 
-if [ "$INSTALL_MODE" = "daemon" ]; then
-    printf '\e[1;31mSwitching to the Daemon-based Installer\e[0m\n'
-    exec "$self/install-multi-user"
-    exit 0
-fi
+    checkEnv() {
+        if [ "$(id -u)" -eq 0 ]; then
+            # TODO: Reword after scripts integration and option switching becomes clear.
+            notice "
 
-if [ "$(id -u)" -eq 0 ]; then
-    printf '\e[1;31mwarning: installing Nix as root is not supported by this script!\e[0m\n'
-fi
+    Install executed for ROOT.
 
-echo "performing a single-user installation of Nix..." >&2
+    Doing classic Linux package manager mode.
+    In Nix this mode is called: single-user mode for root.
+
+    Nix can do multi-user mode, and manage package trees for users independently.
+    "
+        fi
+
+        # In case USER is not set
+        # Example: running inside container
+        if [ -z "$USER" ]; then
+            notice "
+
+    Variable USER is not set in the environment.
+    "
+            USER="$(id -u -n)"    # id is POSIX
+            print "Detected username: $USER"
+        fi
+
+        if [ -z "$HOME" ]; then
+            error "
+
+    Environment variable HOME is not set.
+    "
+        fi
+
+    }
+
+    checkHome() {
+        if [ ! -e "$HOME" ]; then
+            error "
+
+    Home directory $HOME does not exist.
+    "
+        fi
+
+        if [ ! -d "$HOME" ]; then
+            error "
+
+    Home directory $HOME is not a directory.
+    "
+        fi
+
+        if [ ! -w "$HOME" ]; then
+            error "
+
+    Home directory $HOME is not writable for user $USER. No deal.
+    "
+        fi
+
+        if [ "$(ls -ld "$HOME" | awk '{print $3}')" != "$USER" ]; then # POSIX, no `test -O`
+            error "
+
+    Home directory $HOME is not owned by user $USER.
+    If you have legitimate case, file a bug with description.
+    So we can handle your particular case.
+    "
+        fi
+
+        if [ ! -x "$HOME" ]; then
+            error "
+
+    Home directory $HOME is not marked as executable for user $USER,
+    how then we are going to go into it?
+    "
+        fi
+
+    }
+
+    checkDest() {
+     if [ -e "$dest" ]; then
+        # Destination directory exist. Nix is/was installed before, be cautious.
+
+        if [ ! -d "$dest" ]; then
+            error "
+
+    '$dest' exists and it's not a directory.
+    "
+        fi
+
+        if [ ! -w $dest ]; then
+            error "
+
+    Directory '$dest' exists, but is not writable for user '$USER'.
+
+    To enable multi-user support see:
+    http://nixos.org/nix/manual/#ssec-multi-user
+
+    To nevertheless do a single-user install for '$USER':
+    recursively set user '$USER' as owner for '$dest' directory.
```

Comment:
Why not just tell the user exactly the right command based on `chown -R` including sudo if required? (not important)

Reply:
So rookies wound not mindlessly copy-paste the `chown -R` somewhere on public places/on campus.

People should know what they are doing and do things as such themselves.
And it is a rookie question, so:
  * if person does not know - he needs to understand what asked, and then this person understands what he is doing. 
  * if person knows - person writes from memory, or even refreshes from `man` in seconds.

Reply:
I don't agree with this reasoning. Rookies don't have permissions to public folders or campus machines. This is only used for `/nix`. Giving beginners a solid place to start is way nicer than having them solve puzzles.

Reply:
The current web of installers is a software puzzle, and installer in `master` pretty much does not work in many cases.

I gathered a solved here bugreports in the head post.
There should be more now.

For example https://github.com/NixOS/nixpkgs/issues/59370
In Nix reports there are a number of fresh issues.